The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9934324169/job/27438469999) [(plain)](https://github.com/rust-lang/rust/commit/f7c939607f2cef92295a1fe267720547a66891ee/checks/27438469999/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [codegen] tests/codegen/option-as-slice.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-17/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/option-as-slice/option-as-slice.ll" "/checkout/tests/codegen/option-as-slice.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/option-as-slice.rs:55:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/option-as-slice.rs:55:12: error: CHECK: expected string not found in input
 // CHECK: %[[LEN:.+]] = zext{{.+}} i8 %[[TAG]] to i64
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/option-as-slice/option-as-slice.ll:31:52: note: scanning from here
 %0 = load i8, ptr %o, align 1, !range !5, !noundef !4
                                                   ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/option-as-slice/option-as-slice.ll:31:52: note: with "TAG" equal to "0"
 %0 = load i8, ptr %o, align 1, !range !5, !noundef !4
                                                   ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/option-as-slice/option-as-slice.ll:32:2: note: possible intended match here
 %_6 = zext i8 %0 to i64

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/option-as-slice/option-as-slice.ll
Check file: /checkout/tests/codegen/option-as-slice.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'option_as_slice.90015ad8686ed96c-cgu.0' 
            2: source_filename = "option_as_slice.90015ad8686ed96c-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128" 
            4: target triple = "x86_64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
            7: define { ptr, i64 } @u64_opt_as_slice(ptr noalias noundef readonly align 8 dereferenceable(16) %o) unnamed_addr #0 { 
            8: start: 
            9:  %_4 = getelementptr inbounds i8, ptr %o, i64 8 
           10:  %_6 = load i64, ptr %o, align 8, !range !3, !noundef !4 
           11:  %0 = insertvalue { ptr, i64 } poison, ptr %_4, 0 
           12:  %1 = insertvalue { ptr, i64 } %0, i64 %_6, 1 
           13:  ret { ptr, i64 } %1 
           14: } 
           15:  
           16: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
           17: define { ptr, i64 } @nonzero_u64_opt_as_slice(ptr noalias noundef readonly align 8 dereferenceable(8) %o) unnamed_addr #0 { 
           18: start: 
           19:  %0 = load i64, ptr %o, align 8, !noundef !4 
           20:  %1 = icmp ne i64 %0, 0 
           21:  %_4 = zext i1 %1 to i64 
           22:  %2 = insertvalue { ptr, i64 } poison, ptr %o, 0 
           23:  %3 = insertvalue { ptr, i64 } %2, i64 %_4, 1 
           24:  ret { ptr, i64 } %3 
           25: } 
           26:  
           27: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
           28: define { ptr, i64 } @u8_opt_as_slice(ptr noalias noundef readonly align 1 dereferenceable(2) %o) unnamed_addr #0 { 
           29: start: 
           30:  %_4 = getelementptr inbounds i8, ptr %o, i64 1 
           31:  %0 = load i8, ptr %o, align 1, !range !5, !noundef !4 
check:55'0                                                        X~~~ error: no match found
check:55'1                                                             with "TAG" equal to "0"
           32:  %_6 = zext i8 %0 to i64 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
check:55'2      ?                        possible intended match
           33:  %1 = insertvalue { ptr, i64 } poison, ptr %_4, 0 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           34:  %2 = insertvalue { ptr, i64 } %1, i64 %_6, 1 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35:  ret { ptr, i64 } %2 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~
           36: } 
check:55'0     ~~
check:55'0     ~
check:55'0     ~
           38: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:55'0     ~
check:55'0     ~
           40: !llvm.module.flags = !{!0, !1} 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           41: !llvm.ident = !{!2} 
check:55'0     ~~~~~~~~~~~~~~~~~~~~
check:55'0     ~
check:55'0     ~
           43: !0 = !{i32 8, !"PIC Level", i32 2} 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           44: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45: !2 = !{!"rustc version 1.81.0-nightly (06b2c3e4b 2024-07-15)"} 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46: !3 = !{i64 0, i64 2} 
check:55'0     ~~~~~~~~~~~~~~~~~~~~~
           47: !4 = !{} 
check:55'0     ~~~~~~~~~
           48: !5 = !{i8 0, i8 2} 
check:55'0     ~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

--- a/library/core/src/option.rs
+++ b/library/core/src/option.rs
@@ -770,6 +770,13 @@ impl<T> Option<T> {
         }
     }
 
+    #[inline]
+    const fn len(&self) -> usize {
+        // Using the intrinsic avoids emitting a branch to get the 0 or 1.
+        let discriminant: isize = crate::intrinsics::discriminant_value(self);
+        discriminant as usize
+    }
+
     /// Returns a slice of the contained value, if any. If this is `None`, an
     /// empty slice is returned. This can be useful to have a single type of
     /// iterator over an `Option` or slice.
@@ -812,7 +819,7 @@ impl<T> Option<T> {
         unsafe {
             slice::from_raw_parts(
                 (self as *const Self).byte_add(core::mem::offset_of!(Self, Some.0)).cast(),
-                self.is_some() as usize,
+                self.len(),
             )
         }
     }
@@ -869,7 +876,7 @@ impl<T> Option<T> {
         unsafe {
             slice::from_raw_parts_mut(
                 (self as *mut Self).byte_add(core::mem::offset_of!(Self, Some.0)).cast(),
-                self.is_some() as usize,
+                self.len(),
             )
         }
     }
@@ -2242,10 +2249,8 @@ impl<A> Iterator for Item<A> {
 
     #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
-        match self.opt {
-            Some(_) => (1, Some(1)),
-            None => (0, Some(0)),
-        }
+        let len = self.len();
+        (len, Some(len))
     }
 }
 
@@ -2256,7 +2261,12 @@ impl<A> DoubleEndedIterator for Item<A> {
     }
 }
 
-impl<A> ExactSizeIterator for Item<A> {}
+impl<A> ExactSizeIterator for Item<A> {
+    #[inline]
+    fn len(&self) -> usize {
+        self.opt.len()
+    }
+}
 impl<A> FusedIterator for Item<A> {}
 unsafe impl<A> TrustedLen for Item<A> {}
 

--- a/tests/codegen/option-as-slice.rs
+++ b/tests/codegen/option-as-slice.rs
@@ -14,6 +14,14 @@ pub fn u64_opt_as_slice(o: &Option<u64>) -> &[u64] {
     // CHECK-NOT: br
     // CHECK-NOT: switch
     // CHECK-NOT: icmp
+    // CHECK: %[[LEN:.+]] = load i64,{{.+}} !range ![[META_U64:.+]], !noundef
+    // CHECK-NOT: select
+    // CHECK-NOT: br
+    // CHECK-NOT: switch
+    // CHECK-NOT: icmp
+    // CHECK: %[[T0:.+]] = insertvalue { ptr, i64 } poison, ptr %{{.+}}, 0
+    // CHECK-NEXT: %[[T1:.+]] = insertvalue { ptr, i64 } %[[T0]], i64 %[[LEN]], 1
+    // CHECK-NEXT: ret { ptr, i64 } %[[T1]]
     o.as_slice()
 }
 
@@ -25,10 +33,35 @@ pub fn nonzero_u64_opt_as_slice(o: &Option<NonZero<u64>>) -> &[NonZero<u64>] {
     // CHECK-NOT: switch
     // CHECK-NOT: icmp
     // CHECK: %[[NZ:.+]] = icmp ne i64 %{{.+}}, 0
-    // CHECK-NEXT: zext i1 %[[NZ]] to i64
+    // CHECK-NEXT: %[[LEN:.+]] = zext i1 %[[NZ]] to i64
     // CHECK-NOT: select
     // CHECK-NOT: br
     // CHECK-NOT: switch
     // CHECK-NOT: icmp
+    // CHECK: %[[T0:.+]] = insertvalue { ptr, i64 } poison, ptr %o, 0
+    // CHECK-NEXT: %[[T1:.+]] = insertvalue { ptr, i64 } %[[T0]], i64 %[[LEN]], 1
+    // CHECK-NEXT: ret { ptr, i64 } %[[T1]]
     o.as_slice()
 }
+
+// CHECK-LABEL: @u8_opt_as_slice
+#[no_mangle]
+pub fn u8_opt_as_slice(o: &Option<u8>) -> &[u8] {
+    // CHECK-NOT: select
+    // CHECK-NOT: br
+    // CHECK-NOT: switch
+    // CHECK-NOT: icmp
+    // CHECK: %[[TAG:.+]] = load i8,{{.+}} !range ![[META_U8:.+]], !noundef
+    // CHECK: %[[LEN:.+]] = zext{{.*}} i8 %[[TAG]] to i64
+    // CHECK-NOT: select
+    // CHECK-NOT: br
+    // CHECK-NOT: switch
+    // CHECK-NOT: icmp
+    // CHECK: %[[T0:.+]] = insertvalue { ptr, i64 } poison, ptr %{{.+}}, 0
+    // CHECK-NEXT: %[[T1:.+]] = insertvalue { ptr, i64 } %[[T0]], i64 %[[LEN]], 1
+    // CHECK-NEXT: ret { ptr, i64 } %[[T1]]
+    o.as_slice()
+}
+
+// CHECK: ![[META_U64]] = !{i64 0, i64 2}
+// CHECK: ![[META_U8]] = !{i8 0, i8 2}

